File: day2/solution.hs
parseMove str =
  case words str of
    ["forward", m] -> (0, read m :: Int)
    ["down", m] -> (read m :: Int, 0)
    ["up", m] -> (- read m :: Int, 0)
    _ -> error "couldn't parse string"

sumTuples (a, b) (x, y) = (a + x, b + y)

solution = do
  m <- map parseMove . lines <$> readFile "input.txt"
  let (u, v) = foldr sumTuples (0, 0) m
  print $ show $ u * v

File: day2/solution.rs
fn parse_move(line: &str) -> (i32, i32) {
    line.split_once(" ")
        .map(|(dir, mag)| {
            let value = mag.parse::<i32>().unwrap();
            match dir {
                "forward" => (0, value),
                "down" => (value, 0),
                _ => (-value, 0),
            }
        })
        .unwrap()
}

pub fn solution() -> i32 {
    let (x, y) = std::fs::read_to_string("../input.txt")
        .unwrap()
        .lines()
        .fold((0, 0), |(acc1, acc2), line| {
            let (f, s) = parse_move(line);
            (acc1 + f, acc2 + s)
        });
    return x * y;
}
